pkg/virt-launcher/virtwrap/network/ndp/ndp.go

```diff
@@ -0,0 +1,108 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package ndp
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"time"
+
+	"github.com/mdlayher/ndp"
+
+	"kubevirt.io/client-go/log"
+)
+
+const (
+	routerAdvertisementMaxLifetime = 65535 * time.Second // check RFC 4861, section 4.2; 16 bit integer.
+)
+
+type RouterAdvertisementDaemon struct {
+	raOptions []ndp.Option
+	ndpConn   *NDPConnection
+}
+
+func RouterAdvertisementDaemonFromFD(openedFD *os.File, ifaceName string, ipv6CIDR string) (*RouterAdvertisementDaemon, error) {
+	iface, err := net.InterfaceByName(ifaceName)
+	if err != nil {
+		return nil, fmt.Errorf("could not find interface %s: %v", ifaceName, err)
+	}
+
+	ndpConnection, err := importNDPConnection(openedFD, iface)
+	if err != nil {
+		return nil, fmt.Errorf("failed to import the NDP connection from the opened file descriptor")
+	}
+
+	prefix, network, err := net.ParseCIDR(ipv6CIDR)
+	if err != nil {
+		return nil, fmt.Errorf("could not compute prefix / prefix length from %s. Reason: %v", ipv6CIDR, err)
+	}
+	prefixLength, _ := network.Mask.Size()
+	rad := &RouterAdvertisementDaemon{
+		ndpConn:   ndpConnection,
+		raOptions: prepareRAOptions(prefix, uint8(prefixLength)),
+	}
+	return rad, nil
+}
+
+func (rad *RouterAdvertisementDaemon) Serve() error {
+	for {
+		msg, _, err := rad.ndpConn.ReadFrom()
```

Comment:
The srcAddr is ignored here (`_`). I would store the src address and use it in the `WriteTo` instead of sending to `net.IPv6linklocalallnodes`.